### PR TITLE
Refactor various typing related issues

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

This PR has two related aims:
1) Add typing to all `visit_...` and `leave_...` calls in the codebase
2) Fix some outstanding mypy issues
As a result of doing the former the number of mypy issues increased so it seemed logical to take another look at some of the open issues (based on comments meant to ignore mypy). I did not add a changelog contribution.

Before merging there are a small number of calls for which I did not find documentation and therefore I could not add the type annotations. I think it would be best to add those before merging this, to make sure this PR catches all calls.
The missing annotations are for:
~`visit_exec`~
~`visit_default`~
`visit_print`
`visit_repr`
`visit_package`
`visit_instance`
~`visit_project`~
Does any of the maintainers know the nodes passed into these calls? And if so, can you give me a list or add them yourself?

### CODE REVIEW

File: pylint/checkers/typecheck.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/typecheck.py:1842]: We could merge the two ifs here.

File: pylint/extensions/broad_try_clause.py
Comment (Pierre-Sassoulas) [inline pylint/extensions/broad_try_clause.py:13]: Well done, we forget that one apparentely.

File: pylint/checkers/stdlib.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/stdlib.py:447]: Look like a false positive, right ?

**Review by Pierre-Sassoulas — COMMENTED** [review]
Amazing work, there are a lot of removed type ignore ! Maybe we can even add the mypy option for unnecessary type ignore now ? Regarding the missing type I need to check on a computer later.

File: pylint/checkers/stdlib.py
Comment (DanielNoord) [reply pylint/checkers/stdlib.py:447]: Yes, opened #4941

**Comment by DanielNoord** [comment]
I have updated the mypy configuration.

Outstanding issues:

1. `visit_print` comes from `Python3Checker`. Is that checker even relevant after having dropped support for Python < 3.6? I saw a number of disabled messages in our own codebase from this checker which currently do nothing. 
Are we going to keep the checker in the code? And if so, should we remove disabled messages that do nothing currently?
2. `visit_repr`, same as above.
3. `visit_package`, docstring refers to `astroid.Package node` but I don't know what this is.
4. `visit_instance`, this is a `astroid.objects.ExceptionInstance`. However, as `astroid.objects` is not imported into the main `astroid` file we can't access it before runtime and are therefore losing out on IntelliSense. Should we update `astroid` to include this in its main imports and therefore make it accessible to our typing?

**Comment by Pierre-Sassoulas** [comment]
You're right about the python3 checker, I'm going to clean it up, if someone really need to port to python 3 in 2021 that someone can probably suffer to use an older pylint.

File: pylint/checkers/python3.py
Comment (cdce8p) [inline pylint/checkers/python3.py:1045]: Can be removed completely. Parsing legacy code is no longer supported by astroid.
Maybe `visit_print` and `visit_exec` statements should be removed in a separate PR prior to this one?

File: pylint/checkers/python3.py
Comment (cdce8p) [inline pylint/checkers/python3.py:1350]: Should be safe to remove as well

File: pylint/checkers/refactoring/refactoring_checker.py
Comment (cdce8p) [inline pylint/checkers/refactoring/refactoring_checker.py:598]:
```suggestion
    def leave_module(self, _: nodes.Module) -> None:
```

File: pylint/checkers/stdlib.py
Comment (cdce8p) [inline pylint/checkers/stdlib.py:447]:
```suggestion
    def __init__(self, linter: PyLinter = None):  # pylint: disable=super-init-not-called
```
Maybe we should add a link to the issue with the `pylint: disable` statement.
Same for the other time, I think in `exceptions.py`.

File: pylint/checkers/strings.py
Comment (cdce8p) [inline pylint/checkers/strings.py:357]:
```suggestion
                        and arg_type != astroid.Uninferable
```
Don't know, maybe `is not` will work too.

File: pylint/checkers/strings.py
Comment (cdce8p) [inline pylint/checkers/strings.py:403]: Same here

File: pylint/checkers/strings.py
Comment (cdce8p) [inline pylint/checkers/strings.py:381]: We should not use `hasattr`. `isinstance(rhs_tuple, nodes.BaseContainer)` would be better. With that the `if rhs_tuple` check will be redundant too.

File: pylint/checkers/typecheck.py
Comment (cdce8p) [inline pylint/checkers/typecheck.py:1362]:
```suggestion
        parameters: List[List[Union[Tuple[Optional[str], Any], bool]]] = []
```
At some point this should probably be revisited 😅

File: pylint/checkers/typecheck.py
Comment (cdce8p) [inline pylint/checkers/typecheck.py:1644]:
```suggestion
                context_path_names: Iterator[Any] = filter(None, _unflatten(context.path))
```

File: pylint/checkers/typecheck.py
Comment (cdce8p) [inline pylint/checkers/typecheck.py:1802]:
```suggestion
            if inferred and inferred != astroid.Uninferable:
```

File: pylint/checkers/utils.py
Comment (cdce8p) [inline pylint/checkers/utils.py:1169]: I don't really like adding `pylint: disable` here. Better to do `_: nodes.NodeNG` instead.

File: pylint/checkers/utils.py
Comment (cdce8p) [inline pylint/checkers/utils.py:1174]: Same here

File: pylint/checkers/utils.py
Comment (cdce8p) [inline pylint/checkers/utils.py:1252]: Why return `Any`?

File: pylint/checkers/variables.py
Comment (cdce8p) [inline pylint/checkers/variables.py:789]:
```suggestion
    def leave_classdef(self, _: nodes.ClassDef) -> None:
```

File: pylint/checkers/variables.py
Comment (cdce8p) [inline pylint/checkers/variables.py:798]:
```suggestion
    def leave_lambda(self, _: nodes.Lambda) -> None:
```

File: pylint/checkers/variables.py
Comment (cdce8p) [inline pylint/checkers/variables.py:807]:
```suggestion
    def leave_generatorexp(self, _: nodes.GeneratorExp) -> None:
```

File: pylint/checkers/variables.py
Comment (cdce8p) [inline pylint/checkers/variables.py:816]:
```suggestion
    def leave_dictcomp(self, _: nodes.DictComp) -> None:
```

File: pylint/checkers/variables.py
Comment (cdce8p) [inline pylint/checkers/variables.py:825]:
```suggestion
    def leave_setcomp(self, _: nodes.SetComp) -> None:
```

File: pylint/checkers/variables.py
Comment (cdce8p) [inline pylint/checkers/variables.py:1274]:
```suggestion
    def leave_listcomp(self, _: nodes.ListComp) -> None:
```

File: pylint/extensions/check_elif.py
Comment (cdce8p) [inline pylint/extensions/check_elif.py:51]:
```suggestion
    def leave_module(self, _: nodes.Module) -> None:
```

File: pylint/extensions/comparetozero.py
Comment (cdce8p) [inline pylint/extensions/comparetozero.py:60]: Why create a temporary variable?

File: pylint/extensions/docparams.py
Comment (cdce8p) [inline pylint/extensions/docparams.py:616]:
```suggestion
        :type node: nodes.NodeNG
```

File: pylint/extensions/emptystring.py
Comment (cdce8p) [inline pylint/extensions/emptystring.py:58]: ?

File: pylint/extensions/overlapping_exceptions.py
Comment (cdce8p) [inline pylint/extensions/overlapping_exceptions.py:45]:
```suggestion
            handled_in_clause: List[Tuple[Any, Any]] = []
```

File: pylint/extensions/redefined_variable_type.py
Comment (cdce8p) [inline pylint/extensions/redefined_variable_type.py:48]:
```suggestion
    def visit_classdef(self, _: nodes.ClassDef) -> None:
```

File: pylint/extensions/redefined_variable_type.py
Comment (cdce8p) [inline pylint/extensions/redefined_variable_type.py:52]:
```suggestion
    def leave_classdef(self, _: nodes.ClassDef) -> None:
```

File: pylint/extensions/redefined_variable_type.py
Comment (cdce8p) [inline pylint/extensions/redefined_variable_type.py:58]:
```suggestion
    def visit_module(self, _: nodes.Module) -> None:
```

File: pylint/pyreverse/diadefslib.py
Comment (cdce8p) [inline pylint/pyreverse/diadefslib.py:152]:
```suggestion
    def leave_project(self, _: Project) -> Any:
```
?

File: pylint/pyreverse/inspector.py
Comment (cdce8p) [inline pylint/pyreverse/inspector.py:116]:
```suggestion
    def visit_project(self, node:  Project) -> None:
```
?

**Review by cdce8p — DISMISSED** [review]
Awesome work you already did here @DanielNoord 🚀

A few points:
- I don't really like the added `pylint: disable=unused-argument` comments. We should use `_` instead.
- I believe that in the strict mode for mypy even unused arguments should be typed. Those are by far most of my comments.
- `visit_print`, `visit_repr`, and `visit_exec` should get called anymore. Those could be removed. I haven't looked at #4942 just yet, but will do so next.
- We should probably avoid typing generic `list`, `dict`, ... . Better to use `List[Any]`, ... instead.

File: pylint/checkers/utils.py
Comment (DanielNoord) [reply pylint/checkers/utils.py:1252]: Mypy was complaining about `type`. I tried to deduct what class `type` is but got stuck and added `Any`.

File: pylint/extensions/comparetozero.py
Comment (DanielNoord) [reply pylint/extensions/comparetozero.py:60]: Mypy kept complaining about `*ops` not being an `Iterable` but a `List`. I tried other solutions, but this was the only one I could come up with that satisfied both mypy and the tests.

Same for the other one where I do this.

File: pylint/extensions/emptystring.py
Comment (DanielNoord) [reply pylint/extensions/emptystring.py:58]: See above.

File: pylint/pyreverse/diadefslib.py
Comment (DanielNoord) [reply pylint/pyreverse/diadefslib.py:152]: Looks like I missed this one :(

File: pylint/checkers/base.py
Comment (DanielNoord) [reply pylint/checkers/base.py:2024]: Mypy complained about `*args` possibly supplying a value to `prevalent_group` before the second part could do so. This would give multiple values to `prevalent_group`.
This was the suggested change I found online.

File: pylint/checkers/base.py
Comment (DanielNoord) [reply pylint/checkers/base.py]: Done!

File: pylint/checkers/base.py
Comment (DanielNoord) [reply pylint/checkers/base.py]: Done!

File: pylint/checkers/base.py
Comment (DanielNoord) [reply pylint/checkers/base.py]: Done!

File: pylint/checkers/exceptions.py
Comment (DanielNoord) [reply pylint/checkers/exceptions.py]: Done!

File: pylint/checkers/logging.py
Comment (DanielNoord) [reply pylint/checkers/logging.py]: Done!

File: pylint/checkers/python3.py
Comment (DanielNoord) [reply pylint/checkers/python3.py:1043]: See #4942

File: pylint/checkers/python3.py
Comment (DanielNoord) [reply pylint/checkers/python3.py:1348]: See #4942

File: pylint/checkers/strings.py
Comment (DanielNoord) [reply pylint/checkers/strings.py]: Done!

File: pylint/checkers/strings.py
Comment (DanielNoord) [reply pylint/checkers/strings.py]: Done!

File: pylint/checkers/utils.py
Comment (DanielNoord) [reply pylint/checkers/utils.py]: Done!

File: pylint/checkers/utils.py
Comment (DanielNoord) [reply pylint/checkers/utils.py]: Done!

**Comment by DanielNoord** [comment]
Besides open questions from code review, following is needed for this to go forward:
1. Merge of #4942
2. Answer to: `visit_instance`, this is a `astroid.objects.ExceptionInstance`. However, as `astroid.objects` is not imported into the main `astroid` file we can't access it before runtime and are therefore losing out on IntelliSense. Should we update `astroid` to include this in its main imports and therefore make it accessible to our typing?

File: pylint/checkers/typecheck.py
Comment (DanielNoord) [reply pylint/checkers/typecheck.py:1362]: The problem with `Union` is that mypy complains that it can't iterate over `bool`.
Using `Tuple` is impossible Because of call to `append` (mypy complains) and `List` can't have more than 1 argument. Do you have better typing?

File: pylint/pyreverse/inspector.py
Comment (DanielNoord) [inline pylint/pyreverse/inspector.py:83]: Moved this because of typing on line 116. Also added `self.uid = None`

**Review by Pierre-Sassoulas — COMMENTED** [review]
Amazing work done here. I'm pretty sure being able to use proper typing will prevent crashes in the future.

File: pylint/checkers/logging.py
Comment (DanielNoord) [reply pylint/checkers/logging.py:304]: Yeah, it is actually redundant and `is_method_call` makes sure that `func` is a `BoundMethod`. However, this doesn't get recognised by `mypy` or any IDE so we need to add this redundant check here.
Perhaps removing the checks from the return statement in `is_method_call` and doing `if ... return True` might also work. That would remove a duplicate check. However, I am not sure.

File: pylint/checkers/exceptions.py
Comment (DanielNoord) [reply pylint/checkers/exceptions.py]: It did :)

File: doc/how_tos/custom_checkers.rst
Comment (cdce8p) [inline doc/how_tos/custom_checkers.rst:120]: Could you add the import for `PyLinter` in L42 -> `from pylint.lint import PyLinter`?

File: doc/how_tos/custom_checkers.rst
Comment (cdce8p) [inline doc/how_tos/custom_checkers.rst:124]: We need the import from `nodes` as well, in L39

File: pylint/pyreverse/inspector.py
Comment (cdce8p) [reply pylint/pyreverse/inspector.py:83]: Instead of moving it, just use a string as type annotation -> `node: "Project"`

File: pylint/checkers/typecheck.py
Comment (cdce8p) [reply pylint/checkers/typecheck.py:1362]: No without a bigger refactoring. Let's leave it at `List[List[Any]]` for now.

File: pylint/checkers/utils.py
Comment (cdce8p) [reply pylint/checkers/utils.py:1248]: `type` is the generic class type: https://docs.python.org/3/library/typing.html#typing.Type
It should work if you annotate `types = set()` with `Set[type]`.

File: pylint/pyreverse/inspector.py
Comment (DanielNoord) [reply pylint/pyreverse/inspector.py:83]: Don't you think moving it is better? That way we don't define before accessing.

**Comment by DanielNoord** [comment]
I have removed `visit_package`.

Left to-do:

- [ ] Three open questions were have not been resolved.
- [ ] Answer to: visit_instance, this is a astroid.objects.ExceptionInstance. However, as astroid.objects is not imported into the main astroid file we can't access it before runtime and are therefore losing out on IntelliSense. Should we update astroid to include this in its main imports and therefore make it accessible to our typing?

File: pylint/checkers/utils.py
Comment (cdce8p) [inline pylint/checkers/utils.py:1237]:
```suggestion
def node_type(node: nodes.NodeNG) -> Optional[type]:
```
I believe `type` was correct here.

File: pylint/checkers/utils.py
Comment (cdce8p) [inline pylint/checkers/utils.py:1245]:
```suggestion
    types: Set[type] = set()
```

File: pylint/pyreverse/inspector.py
Comment (cdce8p) [inline pylint/pyreverse/inspector.py]: I'm not too familiar with the `pyreverse` code, but even if it isn't called anymore, we shouldn't remove it here. That should be done in a separate PR.

Comment (DanielNoord) [reply pylint/pyreverse/inspector.py]: See #4952

File: pylint/checkers/utils.py
Comment (DanielNoord) [reply pylint/checkers/utils.py:1237]: No worries, good to know all PR's are checked extensively!

**Comment by DanielNoord** [comment]
I have opened https://github.com/PyCQA/astroid/pull/1160 so that the issue with the typing of `visit_instance` can be resolved.
After that merge all that's left is the outstanding question of Pierre and two from Marc, although I think those have been answered and can be resolved!
Rest is good to go I think!

**Comment by cdce8p** [comment]
I did take a look at
```py
# old
ops = list(itertools.chain(*ops))

# new
iter_ops: Iterable[Any] = iter(ops)
ops = list(itertools.chain(*iter_ops))
```
Tbh I think this might be a mypy issue. We can probably leave the fix as you have it now.

--
Due to the large size of this PR, I was wondering if we should merge it sooner rather than later. Can the `visit_instance` issue be resolved in a followup as well? @Pierre-Sassoulas What do you think?

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas already approved my PR for `astroid` so `visit_instance` has been taken care of. I think we can merge this before `astroid` has been released as it does not affect functionality.

**Comment by Pierre-Sassoulas** [comment]
Yeah, this is a huge change and need to be merged relatively quickly or we'll get conflict on a lot of other MR. May as well put it in 2.11.0 as I did not finish #4900 and we can't release it yet 😄

**Comment by DanielNoord** [comment]
Is there anything blocking the merge of this PR? Or do you want to wait on the `astroid` PR?

**Pull request merged**
